src/Compilers/Test/Utilities/CSharp/CompilationTestUtils.cs

```diff
@@ -407,5 +408,40 @@ ExpressionSyntax asExpression(SyntaxNode node)
                 return expr;
             }
         }
+
+        private sealed class TopLevelNullabilityRetreiver : BoundTreeWalker
+        {
+            public Dictionary<SyntaxNode, TypeSymbolWithAnnotations> Map { get; set; }
+
+            protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpression node)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override BoundNode Visit(BoundNode node)
+            {
+                // For the purposes of verifying types, implicit conversions should not win over lower nodes
+                // with the same syntax, as when testing we're usually testing the kind before the conversion,
+                // and we have no way from syntax of getting the correct node.
+                if (node is BoundConversion conv && conv.Syntax == conv.Operand.Syntax && conv.Operand.Kind != BoundKind.Lambda)
+                {
+                    if (node is BoundExpression expr)
+                    {
+                        Map[expr.Syntax] = new TypeWithState(expr.Type, expr.TopLevelNullabilityInfo.FlowState).ToTypeSymbolWithAnnotations();
+                    }
+                    base.Visit(node);
+                }
+                else
+                {
+                    base.Visit(node);
```

Comment:
>base.Visit(node); [](start = 20, length = 17)

is the order of `base.Visit(node)` significant? If not, factor it out?

Reply:
Yes. See the comment above.

---
In reply to: [264941484](https://github.com/dotnet/roslyn/pull/34017#discussion_r264941484) [](ancestors = 264941484)